### PR TITLE
Fixes for the direct and indirect UIInput modes stamping on each othe…

### PR DESCRIPTION
**Pull request by MaxPalmer-UH**
…r, preventing events being raised for the pointer element and the associated unity events

## Summary

Fixes two issues:
1. When both direct and indirect interaction are enabled (InteractionMode==both), the events in the StateChangeActionMap are now raised (i.e. you will hear the sounds when interacting directly and indirectly with the UI elements in the test scene). This should be for both direct and indirect interaction.
2. When both direct and indirect interaction are enabled (InteractionMode==both), the UI elements respond correctly - colour changes, pressed state. These are driven from the Unity events being raised. This should be for both direct and indirect interaction.

Context: events were not being raised correctly (or at all) when the mode is set to both direct and indirect interaction. This is because the logic used to process the tactile and projective states would stamp on top of each others data - e.g.. the pointer states and the time since an interaction with the canvas happened. I've had to make the code aware that the hybrid mode needs special handling of the states and how they are combined.

## Contributor Tasks

_These tasks are for the merge request creator to tick off when creating a merge request._

- [ ] Pair review with a member of the QA team.
- [ ] Add any release testing considerations to the MR for the next release.
- [ ] Check any relevant CHANGELOG files have been updated.
- [ ] Ensure documentation requirements are met e.g., public API is commented.
- [ ] Consider any licensing/other legal implications for this MR e.g. notices required by any new libraries.
- [ ] Add any relevant labels such as `breaking` to this MR.
- [ ] If this MR closes a Jira issue, make sure the fix version on the JIRA issue is set to the correct one.

## Reviewer Tasks

_Add any instructions or tasks for the reviewer such as specific test considerations before this can be merged._

[Use emojis in review threads to communicate intent and help contributors.](https://github.com/ultraleap/UnityPlugin/blob/develop/CONTRIBUTING.md#review-threads)

- [x] Code reviewed.
- [x] Non-code assets e.g. Unity assets/scenes reviewed.
- [ ] Documentation has been reviewed. Includes checking documentation requirements are met and not missing e.g., public API is commented.
- [ ] Checked and agree with release testing considerations added to MR for the next release.

## Closes JIRA Issue

_If this MR closes any JIRA issues list them below in the form `Closes PROJECT-#`_

### CODE REVIEW

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 1]
```diff
@@ -36,14 +36,51 @@ public class PointerElement : MonoBehaviour
...
         private PointerEventData EventData { get; set; }
-        public PointerStates PointerState { get; private set; }
+        public PointerStates PointerStateTactile { get; private set; }
+        public PointerStates PointerStateProjective { get; private set; }
 
-        private PointerStates PrevState { get; set; }
+        /// <summary>
+        /// Gets the aggregated pointer state, based on the interaction mode and the projective and tactile states
+        /// </summary>
+        public PointerStates AggregatePointerState
```
Comment (MaxPalmer-UH) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:45]: I could always change this to be PointerState so the API doesn't change?

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 2]
```diff
@@ -205,16 +308,74 @@ internal void Process(Hand hand, IProjectionOriginProvider projectionOriginProvi
                     throw new ArgumentOutOfRangeException();
             }
 
-            PrevScreenPosition = EventData.position;
-            RaiseEventsForStateChanges();
+            if (EventData != null)
+                PrevScreenPosition = EventData.position;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:312]: can we wrap this in braces please

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 3]
```diff
@@ -205,16 +308,74 @@ internal void Process(Hand hand, IProjectionOriginProvider projectionOriginProvi
...
+                case InteractionCapability.Direct:
+                    RaiseEventsForStateChanges(PointerStateTactile, PrevStateTactile);
+                    break;
+
+                case InteractionCapability.Indirect:
+                    RaiseEventsForStateChanges(PointerStateProjective, PrevStateProjective);
+                    break;
+
+                default:
+                    break;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:336]: unnecessary default?

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 4]
```diff
@@ -205,16 +308,74 @@ internal void Process(Hand hand, IProjectionOriginProvider projectionOriginProvi
...
+
+                case InteractionCapability.Indirect:
+                    if (PointerStateProjective == PointerStates.OffCanvas && PrevStateProjective == PointerStates.OffCanvas)
+                    {
+                        TimeEnteredCanvas = Time.time;
+                    }
+
+                    break;
+                default:
+                    break;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:369]: Unnecessary default?

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 5]
```diff
@@ -483,14 +641,17 @@ private bool GetLookPointerEventData(Hand hand, Vector3 origin, bool forceTipRay
             // Whether or not this will be a raycast through the finger tip
             var tipRaycast = false;
 
+            if (EventData == null)
+                return false;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:645]: can we wrap this is braces please?

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 6]
```diff
@@ -537,8 +698,11 @@ private bool GetLookPointerEventData(Hand hand, Vector3 origin, bool forceTipRay
         /// </summary>
         /// <param name="hand"></param>
         /// <param name="tipRaycastUsed"></param>
-        private void ProcessState(Hand hand, bool tipRaycastUsed)
+        private PointerStates ProcessState(Hand hand, bool tipRaycastUsed, PointerStates PointerState)
         {
+            if (EventData == null)
+                return PointerState;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:704]: can we wrap this is braces please?

File: Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs [hunk 7]
```diff
@@ -588,6 +754,9 @@ private void ProcessState(Hand hand, bool tipRaycastUsed)
         /// <param name="pointData">Pointer event data</param>
         private void UpdatePointer(PointerEventData pointData)
         {
+            if (EventData == null)
+                return;
```
Comment (MattGrayUL) [inline Packages/Tracking Preview/UI Input/Runtime/Scripts/PointerElement.cs:758]: can we wrap this is braces please?

**Review by MattGrayUL — COMMENTED** [review]
minor changes to formatting, otherwise looks good, will approve after the changes :)

**Review by MattGrayUL — APPROVED** [review]
As this is Preview:

Tested in editor, events now fire for both types of interaction via the Preview example scene, provides sounds as a result

**Pull request merged**